hydro_deploy/core/src/azure.rs

```diff
@@ -0,0 +1,570 @@
+use std::collections::HashMap;
+use std::net::SocketAddr;
+use std::path::PathBuf;
+use std::sync::Arc;
+use std::time::Duration;
+
+use anyhow::{Context, Result};
+use async_ssh2_lite::{AsyncSession, SessionConfiguration};
+use async_trait::async_trait;
+use hydroflow_cli_integration::ServerBindConfig;
+use nanoid::nanoid;
+use serde_json::json;
+use tokio::net::TcpStream;
+
+use super::progress::ProgressTracker;
+use super::ssh::LaunchedSSHHost;
+use super::terraform::{TerraformOutput, TerraformProvider, TERRAFORM_ALPHABET};
+use super::util::async_retry;
+use super::{
+    ClientStrategy, Host, HostTargetType, LaunchedHost, ResourceBatch, ResourceResult,
+    ServerStrategy,
+};
+
+// #[derive(Debug)]
+pub struct LaunchedComputeEngine {
+    resource_result: Arc<ResourceResult>,
+    user: String,
+    pub internal_ip: String,
+    pub external_ip: Option<String>,
+}
+
+impl LaunchedComputeEngine {
+    pub fn ssh_key_path(&self) -> PathBuf {
+        self.resource_result
+            .terraform
+            .deployment_folder
+            .as_ref()
+            .unwrap()
+            .path()
+            .join(".ssh")
+            .join("vm_instance_ssh_key_pem")
+    }
+}
+
+#[async_trait]
+impl LaunchedSSHHost for LaunchedComputeEngine {
+    fn server_config(&self, bind_type: &ServerStrategy) -> ServerBindConfig {
+        match bind_type {
+            ServerStrategy::UnixSocket => ServerBindConfig::UnixSocket,
+            ServerStrategy::InternalTcpPort => ServerBindConfig::TcpPort(self.internal_ip.clone()),
+            ServerStrategy::ExternalTcpPort(_) => todo!(),
+            ServerStrategy::Demux(demux) => {
+                let mut config_map = HashMap::new();
+                for (key, underlying) in demux {
+                    config_map.insert(*key, LaunchedSSHHost::server_config(self, underlying));
+                }
+
+                ServerBindConfig::Demux(config_map)
+            }
+            ServerStrategy::Merge(merge) => {
+                let mut configs = vec![];
+                for underlying in merge {
+                    configs.push(LaunchedSSHHost::server_config(self, underlying));
+                }
+
+                ServerBindConfig::Merge(configs)
+            }
+            ServerStrategy::Tagged(underlying, id) => ServerBindConfig::Tagged(
+                Box::new(LaunchedSSHHost::server_config(self, underlying)),
+                *id,
+            ),
+            ServerStrategy::Null => ServerBindConfig::Null,
+        }
+    }
+
+    fn resource_result(&self) -> &Arc<ResourceResult> {
+        &self.resource_result
+    }
+
+    fn ssh_user(&self) -> &str {
+        self.user.as_str()
+    }
+
+    async fn open_ssh_session(&self) -> Result<AsyncSession<TcpStream>> {
```

Comment:
Same here, this is identical to GCP so we should share the logic.

Reply:
I moved the launchedComputeEngine into a shared file